src/monmove.cpp

```diff
@@ -89,6 +119,39 @@ bool monster::can_move_to( const tripoint &p ) const
         }
     }
 
+    if( has_flag( MF_PATH_AVOID_DANGER ) ) {
+        // Don't enter sharp terrain unless tiny, or attacking
+        if( g->m.has_flag( "SHARP", p ) && !( attitude( &( g->u ) ) == MATT_ATTACK ||
+                                              type->size == MS_TINY || has_flag( MF_FLIES ) ) ) {
+            return false;
+        }
+
+        // Don't enter open pits ever unless tiny, can fly or climb well
+        if( !( type->size == MS_TINY || can_climb ) &&
+            ( target == t_pit || target == t_pit_spiked || target == t_pit_glass ) ) {
+            return false;
+        }
+
+        // Don't enter lava ever
+        if( target == t_lava ) {
+            return false;
+        }
+
+        // Don't enter any dangerous fields
+        if( is_dangerous_fields( g->m.field_at( p ) ) ) {
```

Comment:
Part of the difficulty with this is that it requires checks that the two fields be the *same* field, else you run into problems where smoke-bombs suddenly make them willing to throw themselves into acid. It's still doable, but it would be very annoying to do (which is why I mentioned it in the possible later expansions section).